libp2p/peerstore.nim

```diff
@@ -107,36 +129,206 @@ proc set*(addressBook: var AddressBook,
   ## Consider using addressBook.add() as alternative.
   
   addressBook.book[peerId] = addrs
-  addressBook.addrChange(peerId, addressBook.get(peerId)) # Notify clients
+
+  # Notify clients
+  for handler in addressBook.changeHandlers:
+    handler(peerId, addressBook.get(peerId))
+
+#####################
+# Protocol Book API #
+#####################
+
+proc get*(protoBook: ProtoBook,
+          peerId: PeerID): HashSet[string] =
+  ## Get the known protocols of a provided peer.
+  
+  protoBook.book.getOrDefault(peerId,
+                              initHashSet[string]())
+
+proc add*(protoBook: var ProtoBook,
+          peerId: PeerID,
+          protocol: string) = 
+  ## Adds known protocol codec for a given peer. If the peer is not known, 
+  ## it will be set with the provided protocol.
+  
+  protoBook.book.mgetOrPut(peerId,
+                           initHashSet[string]()).incl(protocol)
+  
+  # Notify clients
+  for handler in protoBook.changeHandlers:
+    handler(peerId, protoBook.get(peerId))
+  
+proc delete*(protoBook: var ProtoBook,
+             peerId: PeerID): bool =
+  ## Delete the provided peer from the book.
+  
+  if not protoBook.book.hasKey(peerId):
+    return false
+  else:
+    protoBook.book.del(peerId)
+    return true
+
+proc set*(protoBook: var ProtoBook,
+          peerId: PeerID,
+          protocols: HashSet[string]) =
+  ## Set known protocol codecs for a given peer. This will replace previously
+  ## stored protocols.
+  
+  protoBook.book[peerId] = protocols
+  
+  # Notify clients
+  for handler in protoBook.changeHandlers:
+    handler(peerId, protoBook.get(peerId))
+
+################
+# Key Book API #
+################
+
+proc get*(keyBook: KeyBook,
+          peerId: PeerID): PublicKey =
+  ## Get the known public key of a provided peer.
+  
+  keyBook.book.getOrDefault(peerId,
+                            PublicKey())
+  
+proc delete*(keyBook: var KeyBook,
+             peerId: PeerID): bool =
+  ## Delete the provided peer from the book.
+  
+  if not keyBook.book.hasKey(peerId):
+    return false
+  else:
+    keyBook.book.del(peerId)
+    return true
+
+proc set*(keyBook: var KeyBook,
```

Comment:
Ditto above, where is this called and who is responsible for it? Upon initial connection in nim-libp2p or somewhere else?